pincer/utils/timestamp.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+# MIT License
+#
+# Copyright (c) 2021 Pincer
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files
+# (the "Software"), to deal in the Software without restriction,
+# including without limitation the rights to use, copy, modify, merge,
+# publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+from __future__ import annotations
+from datetime import datetime
+from typing import Optional, TypeVar, Union
+
+DISCORD_EPOCH = 1420070400
+TS = TypeVar("TS", str, datetime, float, int)
+
+
+class Timestamp:
+    """
+    Contains a lot of useful methods for working with timestamps.
+    """
+    def __init__(self, time: Optional[TS] = None):
+        self.__time = Timestamp.parse(time)
+        self.__epoch = Timestamp.to_epoch(self.__time)
+        self.date, self.time = str(self).split()
+
+    @staticmethod
+    def to_epoch(time: datetime) -> int:
+        """
+        Convert a datetime to an epoch.
+
+        :param time:
+            The datetime to convert.
+        """
+        return int(time.timestamp() * 1000)
+
+    @staticmethod
+    def string_to_datetime(string: str) -> datetime:
+        """
+        Convert a string to a datetime object.
+
+        :param string:
+            The string to convert.
+        """
+        return datetime.fromisoformat(string)
+
+    @staticmethod
+    def epoch_to_datetime(epoch: Union[int, float]) -> datetime:
+        """
+        Convert an epoch to a datetime object.
+
+        :param epoch:
+            The epoch to convert to a datetime object.
+        """
+        return datetime.utcfromtimestamp(epoch)
+
+    @staticmethod
+    def parse(time: Optional[TS] = None) -> datetime:
```

Comment:
Add support for type `datetime` as an argument

Reply:
U fixed this urself :)